### PR TITLE
[DRAFT] Point-cloud streaming

### PR DESCRIPTION
**Pull request by mzegar**
1. Created a demo that uses default `PointCloudLayer` to do webrtc streaming 
2. Added demo to the viz app
3. Bringing in the IO module to parse the `.las` file then pass it to the deckgl context

### CODE REVIEW

File: modules/demo/ssr/point-cloud/plugins/point-cloud/index.js [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+/**
+ *
+ * @param {import('fastify').FastifyInstance} fastify
+ */
+function graphSSRClients(fastify) {
+  const graphs  = Object.create(null);
+  const clients = Object.create(null);
+
+  fastify.decorate(module.exports.graphs, graphs);
```
Comment (AjayThorve) [inline modules/demo/ssr/point-cloud/plugins/point-cloud/index.js:32]: I don't think we are using `graphs` in this example, can be removed

File: modules/demo/ssr/point-cloud/plugins/point-cloud/index.js [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+const wrtc            = require('wrtc');
+const {RenderCluster} = require('../../render/cluster');
+
+const {create: shmCreate, detach: shmDetach} = require('shm-typed-array');
+
+module.exports         = graphSSRClients;
+module.exports.graphs  = Symbol('graphs');
```
Comment (AjayThorve) [inline modules/demo/ssr/point-cloud/plugins/point-cloud/index.js:21]: same as previous, `graphs` not used

File: modules/demo/ssr/point-cloud/plugins/point-cloud/index.js [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+  function onClose(sock, peer) {
+    const [stream] = peer?.streams || [];
+    if (stream) { delete clients[stream.id]; }
+    const {g: graphId = 'default'} = sock?.handshake?.query || {};
+    if (graphId in graphs) {
+      if ((graphs[graphId].refCount -= 1) === 0) {  //
+        delete graphs[graphId];
+      }
+    }
```
Comment (AjayThorve) [inline modules/demo/ssr/point-cloud/plugins/point-cloud/index.js:89]:
```suggestion
```

File: modules/demo/ssr/point-cloud/plugins/point-cloud/index.js [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+  }
+}
+
+function layoutAndRenderGraphs(clients) {
```
Comment (AjayThorve) [inline modules/demo/ssr/point-cloud/plugins/point-cloud/index.js:93]: Could be renamed to PointCloud instead? 
```suggestion
function layoutAndRenderPointCloud(clients) {
```

File: modules/demo/ssr/point-cloud/render/render.js [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    const onAnimationFrameRequested = immediateAnimationFrame(this);
+    const jsdom                     = new RapidsJSDOM({module, onAnimationFrameRequested});
+
+    const {deck, render} = jsdom.window.evalFn(makeDeck);
+
+    this.deck    = deck;
+    this.jsdom   = jsdom;
+    this._render = render;
+  }
+  async render(props = {}, graph = {}, state = {}, events = [], frame = 0) {
```
Comment (AjayThorve) [inline modules/demo/ssr/point-cloud/render/render.js:29]:
```suggestion
  async render(props = {}, state = {}, events = [], frame = 0) {
```

File: modules/demo/ssr/point-cloud/render/render.js [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+  registerLoaders(LASLoader);
+
+  // Data source: kaarta.com
+  const LAZ_SAMPLE =
+    'https://raw.githubusercontent.com/uber-common/deck.gl-data/master/examples/point-cloud-laz/indoor.0.1.laz';
+
+  const transitionInterpolator = new LinearInterpolator(['rotationOrbit']);
+
+  const makeLayers = (deck, graph = null) => {
```
Comment (AjayThorve) [inline modules/demo/ssr/point-cloud/render/render.js:107]: I think renaming `graph` to `pointCloud` might lead to less confusion?
```suggestion
  const makeLayers = (deck, pointCloud = null) => {
```

File: modules/demo/ssr/point-cloud/render/worker.js [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+const {fromEvent, EMPTY}                  = require('rxjs');
+const {tap, groupBy, mergeMap, concatMap} = require('rxjs/operators');
+
+const {Renderer} = require('./render');
+const renderer   = new Renderer();
+
+function render({props = {}, graph = {}, state = {}, event = [], frame = 0}) {
+  return renderer.render(props, graph, state, event, frame);
+}
```
Comment (AjayThorve) [inline modules/demo/ssr/point-cloud/render/worker.js:23]:
```suggestion
function render({props = {}, state = {}, event = [], frame = 0}) {
  return renderer.render(props, state, event, frame);
}
```

**Review by AjayThorve — CHANGES_REQUESTED** [review]
looks good, just a few minor comments

File: modules/demo/ssr/point-cloud/plugins/point-cloud/index.js [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+const {create: shmCreate, detach: shmDetach} = require('shm-typed-array');
+
+module.exports         = graphSSRClients;
+module.exports.clients = Symbol('clients');
+
+/**
+ *
+ * @param {import('fastify').FastifyInstance} fastify
+ */
+function graphSSRClients(fastify) {
```
Comment (mzegar) [inline modules/demo/ssr/point-cloud/plugins/point-cloud/index.js:27]: Should this be renamed as well?

**Comment by mzegar** [comment]
I can throw up an issue to combine some of this duplicate code since I think I'm transitioning over to the thrust stuff next week.

**Comment by trxcllnt** [comment]
Closing in favor of https://github.com/rapidsai/node/pull/348

**Pull request closed without merging**